### ISSUE TITLE
Device streams failing to maintain RTSP stream connection

### ISSUE THREAD

**Issue by langebo** [Azure/azure-iot-sdk-csharp#1645]
# Context
- **OS, version, SKU and CPU architecture used:** macOS Catalina on the service side / Ubuntu 18.04 (amd64) on the device side, Azure IoT Hub (F1 SKU), Azure COntainer Registry (Basic SKU), Azure LogAnalytics 
- **Application's .NET Target Framework :** netcoreapp3.1 
- **Device:** Laptop (MacBook with Ubuntu VM in Parallels Desktop)
- **SDK version used:** Microsoft.Azure.Devices 1.27.0-preview-004 (service) / Microsoft.Azure.Devices.Client 1.32.0-preview-001 (device) 

## Description of the issue
I am currently evaluating the device streams feature for a use case, where we want to stream live video from an RTSP stream that is within the device network. I followed the device stream quickstarts for SSH/RDP. 

The error that I am receiving is that the RTSP client (VLC Media Player in this case) initiates the connection, exchanges some information (it detects the length of the video that is being requested from the RTSP server on the device), but then the service application throws following exception:

```
Unable to read data from the transport connection: Connection reset by peer.
at DeviceProxyService.Streaming.StreamingClient.HandleOutgoingDataAsync(NetworkStream localStream, ClientWebSocket remoteStream, CancellationToken cancellationToken) in /Users/blangrbook/dev/work/scanblue/rtsp-device-streams/service/Streaming/StreamingClient.cs:line 84
```

and the device module throws the following exception:

```
The remote party closed the WebSocket connection without completing the close handshake.
   at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
   at DeviceProxy.Streaming.StreamingClient.HandleIncomingDataAsync(NetworkStream localStream, ClientWebSocket remoteStream, CancellationToken cancellationToken) in /app/Streaming/StreamingClient.cs:line 54
Unhandled exception. System.Net.WebSockets.WebSocketException (0x80004005): The WebSocket is in an invalid state ('Aborted') for this operation. Valid states are: 'Open, CloseReceived, CloseSent'
   at System.Net.WebSockets.WebSocketValidate.ThrowIfInvalidState(WebSocketState currentState, Boolean isDisposed, WebSocketState[] validStates)
   at System.Net.WebSockets.ManagedWebSocket.CloseAsync(WebSocketCloseStatus closeStatus, String statusDescription, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at DeviceProxy.Streaming.StreamingClient.RunProxyAsync(CancellationToken cancellationToken) in /app/Streaming/StreamingClient.cs:line 43
   at DeviceProxy.Program.Init(CancellationToken cancellationToken) in /app/Program.cs:line 32
   at DeviceProxy.Program.Main(String[] args) in /app/Program.cs:line 20
``` 

What am I doing wrong? Is transferring video stream data this way the wrong use case for device streams? That would be sad, since we really need the encrypted and tunnelled proxy functionality.

When I directly call the RTSP module's stream via VLC it works fine.

## Code sample exhibiting the issue
For you to be able to reproduce the issue, I pushed our evaluation sample here:
https://github.com/langebo/rtsp-device-streams

For the RTSP simulator module I created an image according to your live-video-analytics repo here:
https://github.com/Azure/live-video-analytics/tree/master/utilities/rtspsim-live555

On the device side I simply mount this video file into the media folder of the RTSP module:
https://filesamples.com/samples/video/mkv/sample_640x360.mkv

In order to run the project you will need to configure the service app within its Program.cs with:
- iot hub service connection string 
- device id
- local port

For the module you need to provide:
- container registry credentials (i ommited the .env file in the repo)
- iot hub device connection string (eg. as an env variable of the proxy module)

**Comment by langebo** [comment]
This can be closed. The VLC player seems to connect to the socket to acquire media information, then disconnects and connects again for the actual playback. I simply adjusted the application and voilà we can view RTSP streams from our devices. 

I hope I didn't waste anyones time. Keep up the great work.

**Comment by az-iot-builder-01** [comment]
@langebo, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by dme-development** [comment]
> This can be closed. The VLC player seems to connect to the socket to acquire media information, then disconnects and connects again for the actual playback. I simply adjusted the application and voilà we can view RTSP streams from our devices.
> 
> I hope I didn't waste anyones time. Keep up the great work.

@langebo : I see you got it working in the end. I'm exploring the same path as you and facing the same difficulties. Would you mind sharing your use case sample project?

**Comment by langebo** [comment]
@dme-development Sorry for responding so late. In case you are still facing the problem, I simply changed the RunProxyAsync() method in the Proxy.cs on the module side to perform the WaitForDeviceStreamRequestAsync() method inside a loop. On the service side I changed ConnectProxyAsync() in the ProxyClient.cs to also perform AcceptTcpClientAsync() in a loop. 

This way after the first connection (querying metadata) is closed, both sides will be waiting for the second connection, which will actually perform the media playback.
